### PR TITLE
Adds Gloas fork support to the block_sim simulator

### PR DESCRIPTION
**Pull request by Tomi-3-0**

### CODE REVIEW

File: research/block_sim.nim [hunk 1]
```diff
@@ -252,8 +259,73 @@ cli do(
...
+    if slot == GENESIS_SLOT:
+      return
+
+    let previousSlot = slot - 1
+
+    dag.withUpdatedState(tmpState[],
+        dag.head.atSlot(slot).toBlockSlotId.expect("not nil")):
+      let
+        fork = getStateField(updatedState, fork)
+        genesis_validators_root = getStateField(updatedState, genesis_validators_root)
```
Comment (tersec) [inline research/block_sim.nim]: There's a DAG helper function for this:
https://github.com/status-im/nimbus-eth2/blob/ce4689f78b249bbb9e9ea4b55e5d6dc559fcffbb/beacon_chain/consensus_object_pools/blockchain_dag.nim#L1002-L1003

File: research/block_sim.nim [hunk 2]
```diff
@@ -252,8 +259,73 @@ cli do(
...
+      signature: pendingPayloadAttestations[0].signature # TODO: aggregate properly
+    )
+
+    pendingPayloadAttestations.setLen(0)
+
+    res
+
+  let
+    blockRatio = blockRatio # can't find in proposeBlock otherwise (?)
+    builderRatio = builderRatio
```
Comment (tersec) [inline research/block_sim.nim]: Both of these are presumably due to:
- https://github.com/nim-lang/Nim/issues/20811

Useful to link this to track if it's fixed.

**Comment by tersec** [comment]
https://github.com/status-im/nimbus-eth2/blob/f9eb14752e64e1a636ffede69df77bea58421c3b/research/simutils.nim#L57-L65 controls the fork schedule for this.

File: research/block_sim.nim (hunk 1 above)
Comment (tersec) [reply research/block_sim.nim]: https://github.com/status-im/nimbus-eth2/pull/7840/changes/666bc8b0e3ee701402c5fc02d713aa71d8c2bb71

File: research/block_sim.nim (hunk 2 above)
Comment (tersec) [reply research/block_sim.nim]: https://github.com/status-im/nimbus-eth2/pull/7840/changes/666bc8b0e3ee701402c5fc02d713aa71d8c2bb71

**Comment by tersec** [comment]
We currently test the engine API conversions (well, through Deneb, apparently the tests stopped being updated at that point) via https://github.com/status-im/nimbus-eth2/blob/unstable/tests/test_engine_api_conversions.nim

The idea is to mix:
- random values
- boundary values lists with 0 or 1, or max items

in currently 25 examples per fork, but the exact number isn't critical, to verify that the EL encoding/decoding round-trip, i.e. the core loop is
https://github.com/status-im/nimbus-eth2/blob/83eb032bc34b50a67cb38e58b13a7cbadb17c332/tests/test_engine_api_conversions.nim#L1604-L1607

**Comment by Tomi-3-0** [comment]
> We currently test the engine API conversions (well, through Deneb, apparently the tests stopped being updated at that point) via https://github.com/status-im/nimbus-eth2/blob/unstable/tests/test_engine_api_conversions.nim
> 
> The idea is to mix:
> 
> * random values
> * boundary values lists with 0 or 1, or max items
> 
> in currently 25 examples per fork, but the exact number isn't critical, to verify that the EL encoding/decoding round-trip, i.e. the core loop is
> 
> https://github.com/status-im/nimbus-eth2/blob/83eb032bc34b50a67cb38e58b13a7cbadb17c332/tests/test_engine_api_conversions.nim#L1604-L1607

okay

**Comment by Tomi-3-0** [comment]
> https://github.com/status-im/nimbus-eth2/blob/f9eb14752e64e1a636ffede69df77bea58421c3b/research/simutils.nim#L57-L65
> 
> controls the fork schedule for this.

https://github.com/status-im/nimbus-eth2/pull/7840/commits/e2814e78bd102887e3474a43ce5b42f5a7cb1983

**Comment by tersec** [comment]
```
The following files do not have an up-to-date copyright year:
- tests/test_engine_api_conversions.nim
Error: Process completed with exit code 2.
```

File: beacon_chain/validators/block_payloads.nim [hunk 3]
```diff
@@ -173,7 +174,8 @@ func decodePayloadRequests(
   ok default(ExecutionRequests)
 
 func decodePayloadRequests(
-    eps: electra.ExecutionPayloadForSigning | fulu.ExecutionPayloadForSigning
+    eps: electra.ExecutionPayloadForSigning | fulu.ExecutionPayloadForSigning |
+    gloas.ExecutionPayloadForSigning
```
Comment (tersec) [inline beacon_chain/validators/block_payloads.nim]: `gloas.` should be indented (generally nice to align it with `electra.`, but it definitely shouldn't be on the same indentation level as `eps:`)

File: beacon_chain/el/el_manager.nim [hunk 4]
```diff
@@ -435,9 +437,18 @@ proc getPayloadFromSingleEL(
...
+  elif GetPayloadResponseType is engine_api.GetPayloadV3Response:
+    return await engine_getPayloadV3(rpcClient, payloadId)
+  elif GetPayloadResponseType is engine_api.GetPayloadV4Response:
+    return await engine_getPayloadV4(rpcClient, payloadId)
+  elif GetPayloadResponseType is engine_api.GetPayloadV5Response:
+    return await engine_getPayloadV5(rpcClient, payloadId)
+  elif GetPayloadResponseType is engine_api.GetPayloadV6Response:
+    return await engine_getPayloadV6(rpcClient, payloadId)
   else:
-    return await engine_api.getPayload(
```
Comment (tersec) [inline beacon_chain/el/el_manager.nim:439]: With
- https://github.com/status-im/nimbus-eth2/pull/7859

might be able to use this for everything.

**Comment by tersec** [comment]
As a general point, it's unclear why getting `block_sim` to work in Gloas should require, e.g., any engine API changes at all.

**Comment by Tomi-3-0** [comment]
> As a general point, it's unclear why getting `block_sim` to work in Gloas should require, e.g., any engine API changes at all.

This could actully be two seperate PR's. But since block_sim uses the same `makeBeaconBlock` that production uses, I just thought it made sense to update the underlying types and infrastructure here

File: beacon_chain/spec/state_transition.nim [hunk 5]
```diff
@@ -434,6 +434,61 @@ proc makeBeaconBlockWithRewards*(
...
+    randao_reveal: ValidatorSig,
+    eth1_data: Eth1Data,
+    graffiti: GraffitiBytes,
+    attestations: seq[electra.Attestation],
+    deposits: seq[Deposit],
+    validator_changes: BeaconBlockValidatorChanges,
+    sync_aggregate: SyncAggregate,
+    signed_execution_payload_bid: SignedExecutionPayloadBid,
+    verificationFlags: UpdateFlags,
+    payload_attestations: seq[PayloadAttestation],
```
Comment (tersec) [inline beacon_chain/spec/state_transition.nim]: It avoids repetition to add this `payload_attestations` parameter to the existing `makeBeaconBlockWithRewards` and then, if pre-Gloas, pass an empty sequence (ideally even a `const`/`static` version, so it's as inexpensive as it can be at runtime). Otherwise one ends up with a badly scaling setup where each fork which adds a new parameter creates a new `makeBeaconBlockWithRewards` and a new `makeBeaconBlock`. It's an intentional design to centralize all that in one `makeBeaconBlockWithRewards` and essentially one `makeBeaconBlock`.

Comment (tersec) [reply beacon_chain/spec/state_transition.nim]: https://github.com/status-im/nimbus-eth2/pull/7840/commits/922dbdd070321a0009d68853fbfba509bb66a694

File: beacon_chain/el/el_manager.nim (hunk 4 above)
Comment (tersec) [reply beacon_chain/el/el_manager.nim:439]: Not relevant anymore now that it doesn't include engine API changes.

File: beacon_chain/validators/block_payloads.nim (hunk 3 above)
Comment (tersec) [reply beacon_chain/validators/block_payloads.nim]: Not relevant anymore since the engine API part isn't part of this PR.

Comment (Tomi-3-0) [reply beacon_chain/validators/block_payloads.nim]: This has been removed

File: beacon_chain/spec/state_transition.nim [hunk 6]
```diff
@@ -367,6 +367,9 @@ proc makeBeaconBlockWithRewards*(
     verificationFlags: UpdateFlags,
     kzg_commitments: KzgCommitments,
     execution_requests: ExecutionRequests,
+    signed_execution_payload_bid:
+      SignedExecutionPayloadBid = default(SignedExecutionPayloadBid),
+    payload_attestations: seq[PayloadAttestation] = @[]
```
Comment (tersec) [inline beacon_chain/spec/state_transition.nim]: It might be somewhat nicer not to have these as default parameters, because it makes it more error-prone to find all callers which might need updating

Comment (tersec) [reply beacon_chain/spec/state_transition.nim]: https://github.com/status-im/nimbus-eth2/pull/7840/commits/f5008d3ab16359c0d03f2599b4d2a40028a698a3
